SwiftCompilerSources/Sources/Optimizer/ModulePasses/MandatoryPerformanceOptimizations.swift

```diff
@@ -123,9 +132,66 @@ private func shouldInline(apply: FullApplySite, callee: Function) -> Bool {
     // Force inlining them in global initializers so that it's possible to statically initialize the global.
     return true
   }
+  if apply.parentFunction.isGlobalInitOnceFunction,
+      let global = getGlobal(of: apply.parentFunction),
+      global.mustBeInitializedStatically,
+      let applyInst = apply as? ApplyInst,
+      applyInst.isStored(to: global) {
+    // Inline calls in global initializers that end up contributing to the final value of the global
+    return true
+  }
   return false
 }
 
+private extension Value {
+  func isStored(to global: GlobalVariable) -> Bool {
+    var singleUseValue: any Value = self
+    var goesThroughAMemberInit = false
+    while true {
+      switch singleUseValue {
+      case is StructInst,
+          is TupleInst,
+          is EnumInst:
+        goesThroughAMemberInit = true
+        break
+      default:
+        break
+      }
+      guard let use = singleUseValue.uses.singleUse else {
+        return false
+      }
+      
+      if let si = use.instruction as? StoreInst {
+        guard let storeDestination = si.destination as? GlobalAddrInst else {
+          return false
+        }
+        
+        return /*goesThroughAMemberInit &&*/ storeDestination.global == global
```

Comment:
The "require that the def-use chain goes through a struct/tuple/enum inst" (goesThroughAMemberInit) part unfortunately seems to reject some simple situations where we would want the inlining to happen, e.g. this SIL:
```
// one-time initialization function for g_MyStruct2
sil private [global_init_once_fn] [ossa] @$s4null11g_MyStruct2_WZ : $@convention(c) (Builtin.RawPointer) -> () {
bb0(%0 : $Builtin.RawPointer):
  alloc_global @$s4null11g_MyStruct2AA0bC0Vvp     // id: %1
  %2 = global_addr @$s4null11g_MyStruct2AA0bC0Vvp : $*MyStruct2 // user: %8
  %3 = metatype $@thin MyStruct2.Type             // user: %7
  %4 = integer_literal $Builtin.Int64, 42         // user: %5
  %5 = struct $Int (%4 : $Builtin.Int64)          // user: %7
  // function_ref MyStruct2.init(a:)
  %6 = function_ref @$s4null9MyStruct2V1aACSi_tcfC : $@convention(method) (Int, @thin MyStruct2.Type) -> MyStruct2 // user: %7
  %7 = apply %6(%5, %3) : $@convention(method) (Int, @thin MyStruct2.Type) -> MyStruct2 // user: %8
  store %7 to [trivial] %2 : $*MyStruct2          // id: %8
  %9 = tuple ()                                   // user: %10
  return %9 : $()                                 // id: %10
} // end sil function '$s4null11g_MyStruct2_WZ'
```
Ideas?

Reply:
You are right.
Let's drop that requirement for now. I have another idea to prevent infinite inlining, but that's a bit more complicate.

Reply:
PR is redone to use the SmallProjectionPath-based tracking of already inlined functions.